sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py

```diff
@@ -132,21 +132,39 @@ def __init__(
         self._populate_attributes(**kwargs)
 
     def __iter__(self):
-        return self
+        return self._iter_contextual_wrapper()
+
+    def _iter_contextual_wrapper(self, max_wait_time=None):
+        original_timeout = None
+        while True:
+            # This is not threadsafe, but gives us a way to handle if someone passes
+            # different max_wait_times to different iterators and uses them in concert.
+            if max_wait_time:
+                original_timeout = self._handler._timeout
+                self._handler._timeout = max_wait_time * 1000
+            try:
+                yield next(self)
+            except StopIteration:
+                break
+            finally:
+                if original_timeout:
+                    self._handler._timeout = original_timeout
 
     def __next__(self):
         self._check_live()
         while True:
             try:
                 return self._do_retryable_operation(self._iter_next)
             except StopIteration:
-                self.close()
+                self._message_iter = None
                 raise
 
     next = __next__  # for python2.7
 
     def _iter_next(self):
         self._open()
+        if not self._message_iter:
+            self._message_iter = self._handler.receive_messages_iter()  # pylint: disable=attribute-defined-outside-init
```

Comment:
Does this act to "restart" the iterator if it has been closed?

Reply:
While you're probably saying the same thing I'm about to, I'd pedantically call it "reconnecting" the iterator, to convey that the iter is basically a.... conduit, between the underlying amqp receiver and the python layer.  (This may be total semantic nitpicking but figured I'd share as it was a useful mental model when I was wrapping my head around this)